File: erlang/pe017.erl
-module(pe017).
-export([start/0]).

thoucount() -> 11. % "one thousand"
hcount() -> 7. % "hundred" 
handcount() -> 10. % "hundred and "
units() -> ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"].
tens() -> ["ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen", "twenty", "twentyone", "twentytwo", "twentythree", "twentyfour", "twentyfive", "twentysix", "twentyseven", "twentyeight", "twentynine", "thirty", "thirtyone", "thirtytwo", "thirtythree", "thirtyfour", "thirtyfive", "thirtysix", "thirtyseven", "thirtyeight", "thirtynine", "forty", "fortyone", "fortytwo", "fortythree", "fortyfour", "fortyfive", "fortysix", "fortyseven", "fortyeight", "fortynine", "fifty", "fiftyone", "fiftytwo", "fiftythree", "fiftyfour", "fiftyfive", "fiftysix", "fiftyseven", "fiftyeight", "fiftynine", "sixty", "sixtyone", "sixtytwo", "sixtythree", "sixtyfour", "sixtyfive", "sixtysix", "sixtyseven", "sixtyeight", "sixtynine", "seventy", "seventyone", "seventytwo", "seventythree", "seventyfour", "seventyfive", "seventysix", "seventyseven", "seventyeight", "seventynine", "eighty", "eightyone", "eightytwo", "eightythree", "eightyfour", "eightyfive", "eightysix", "eightyseven", "eightyeight", "eightynine", "ninety", "ninetyone", "ninetytwo", "ninetythree", "ninetyfour", "ninetyfive", "ninetysix", "ninetyseven", "ninetyeight", "ninetynine"].

start() ->
  UC = lcount(units(), 0),
  TC = lcount(tens(), 0),
  HC = clcount(units(), hcount(), 0),
  HAUC = clucount(units(), handcount(), 0),
  HATC = clucount(tens(), handcount(), 0),
  io:format("~w~n", [thoucount()+HATC+HAUC+HC+TC+UC]).

lcount([], Result) ->
  Result;
  
lcount([H|T], Result) ->
  lcount(T, length(H) + Result).

clcount([], _C, Result) ->
  Result;
  
clcount([H|T], C, Result) ->
  clcount(T, C, length(H) + C + Result).

clucount([], _C, Result) ->
  Result;
  
% need to find out how to pass through 2 lists and loop over them nested
% don't really like to hardcode units() here...

clucount([H|T], C, Result) ->
  X = clcount(units(), length(H) + C, 0),
  clucount(T, C, X + Result).

File: forth/pe017.fs
#! /usr/bin/gforth

: letter_count (  count ptr len -- numletters )
0 do
  dup
  i + c@ 32 = 0=
  if 
    swap 1+ swap
  then
loop
drop
;

: "units"  c"   one  twothree four five  sixseveneight nine" ;
: unitsl   5 *  "units" 1+ +  5 letter_count ;

: "tens" c"       ten   eleven   twelve thirteen fourteen  fifteen  sixteenseventeen eighteen nineteen" ;
: tensl   9 *  "tens" 1+ +  9 letter_count ;

: "twenties" c"      twenty  twentyone  twentytwotwentythree twentyfour twentyfive  twentysixtwentyseventwentyeight twentynine" ;
: twentiesl   11 *  "twenties" 1+ +  11 letter_count ;

: "thirties" c"      thirty  thirtyone  thirtytwothirtythree thirtyfour thirtyfive  thirtysixthirtyseventhirtyeight thirtynine" ;
: thirtiesl   11 *  "thirties" 1+ +  11 letter_count ;

: "forties" c"      forty  fortyone  fortytwofortythree fortyfour fortyfive  fortysixfortysevenfortyeight fortynine" ; 
: fortiesl   10 *  "forties" 1+ +  10 letter_count ;

: "fifties" c"      fifty  fiftyone  fiftytwofiftythree fiftyfour fiftyfive  fiftysixfiftysevenfiftyeight fiftynine" ;
: fiftiesl   10 *  "fifties" 1+ +  10 letter_count ;

: "sixties" c"      sixty  sixtyone  sixtytwosixtythree sixtyfour sixtyfive  sixtysixsixtysevensixtyeight sixtynine" ;
: sixtiesl   10 *  "sixties" 1+ +  10 letter_count ;

: "seventies" c"      seventy  seventyone  seventytwoseventythree seventyfour seventyfive  seventysixseventysevenseventyeight seventynine" ;
: seventiesl   12 *  "seventies" 1+ +  12 letter_count ;

: "eighties" c"      eighty  eightyone  eightytwoeightythree eightyfour eightyfive  eightysixeightyseveneightyeight eightynine" ;
: eightiesl   11 *  "eighties" 1+ +  11 letter_count ;

: "nineties" c"      ninety  ninetyone  ninetytwoninetythree ninetyfour ninetyfive  ninetysixninetysevenninetyeight ninetynine" ;
: ninetiesl   11 *  "nineties" 1+ +  11 letter_count ;

: hundredl 7 ;
: hundredandl 10 ;
: onethousandl 11 ;

: firsthundred ( sum -- sum )
9 0 do
  0 i unitsl +
loop 
10 0 do
  0 i tensl +
loop 
10 0 do
  0 i twentiesl +
loop 
10 0 do
  0 i thirtiesl +
loop 
10 0 do
  0 i fortiesl +
loop 
10 0 do
  0 i fiftiesl +
loop 
10 0 do
  0 i sixtiesl +
loop 
10 0 do
  0 i seventiesl +
loop 
10 0 do
  0 i eightiesl +
loop 
10 0 do
  0 i ninetiesl +
loop 
;

: hundred ( sum -- sum )
dup rot rot
9 0 do
  0 swap unitsl + 
  hundredandl + 
  0 i unitsl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i tensl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i twentiesl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i thirtiesl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i fortiesl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i fiftiesl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i sixtiesl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i seventiesl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i eightiesl +
  swap dup rot rot
loop 
10 0 do
  0 swap unitsl + hundredandl + 0 i ninetiesl +
  swap dup rot rot
loop 
drop swap drop
;

: pe017 ( -- )
0
firsthundred
9 0 do
  0 i unitsl + hundredl +
  i hundred
loop
onethousandl +
. cr
;

pe017
bye
